FSS purge list reference model initial delivery

jni2000 opened the pull request:
Note to reviewers
This is the very initial submission of the FSS purge list code. Please take a review and give your comments before we go to far if there are any concerns. Also please note that:
1.	This model uses the Objects that are defined in PR#404.
2.	The main function for this model is fssPurgeModel()

sbdt-google commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,357 @@
+           cbsd['grantRequests'].remove(grant_items)
+
+        # If no grants are remaining then remove the CBSD.
+        if len(cbsd['grantRequests']) == 0:
```
prefer: if not cbsd['grantRequests']:

jni2000 replied:
Modified as   suggested.

sbdt-google commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,357 @@
+  """ Convert AMSL height to AGL height."""
+
+  altitude_cbsd = terrain.GetTerrainElevation(latitude, longitude)
+  return height_amsl - altitude_cbsd
```
What to do if the resulting height_agl is negative ?? For example if inconsistent input versus the terrain database.
Question for Kate and others

jni2000 replied:
We will update when answer from working group is available.

sbdt-google commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,357 @@
+def convertAmslToAgl(latitude, longitude, height_amsl):
+  """ Convert AMSL height to AGL height."""
+
+  altitude_cbsd = terrain.GetTerrainElevation(latitude, longitude)
```
`terrain` is a module. You need to call the function on a terrain driver.
Currently there is only one terrain driver initialized in the wf_itm module. I should perhaps move it to the terrain.py module.
For now you can change into:
```
altitude_cbsd = wf_itm.terrainDriver.GetTerrainElevation(latitude, longitude)
```

jni2000 replied:
Corrected.

sbdt-google commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,357 @@
+      performPurge(neighbouring_cbsds, fss_entity, fss_oobe_margin)
+
+      # Removing the oobe_interference parameter added to cbsd objects
+      for cbsd in neighbouring_cbsds:
```
The adding of this field is done inside performPurge(), so it should be cleaned up inside this function also.

jni2000 replied:
Moved inside   perform_purge method.

jni2000 commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,357 @@
+import numpy as np
+import sys
+
+sys.path.append('../..')
```
Thanks.   This line has been removed.

jni2000 commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,357 @@
+FREQ_PROP_MODEL = 3625.0
+
+# Reference Channel bandwidth for FSS OOBE calculation
+REF_BW = 5000000.0
```
Modified as   suggested.

jni2000 commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,357 @@
+
+# Define FSS information tuple containing the parameters necessary for antenna
+# and propagation calculation as listed in the test spec
+FssProtectionPoint = namedtuple('FssEntity',
```
Modified as   suggested.

jni2000 commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,357 @@
+  calculateOobeInterference(cbsds, fss_entity)
+
+  # Sort the CBSDs based on the calculated interference value.
+  sfss = sorted(cbsds, key=lambda cbsd: cbsd['oobe_interference'])
```
The   name sfss is misleading. It was used to keep it consistent with the algorithm   provided in R2-SGN-29 for easy reference.          I agree that it is rather good to have it more readable and meaningful.   sfss has been changed to sorted_cbsds.

jni2000 commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,357 @@
+  # Sort the CBSDs based on the calculated interference value.
+  sfss = sorted(cbsds, key=lambda cbsd: cbsd['oobe_interference'])
+  # Calcualte the threshold value in dBm for the OOBE interference
+  oobe_threshold_value = (-1 * 129) + 10 * (np.log10(REF_BW/ONE_MHZ) - fss_oobe_margin)
```
Changed.

jni2000 commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,357 @@
+  are removed from the CBSDs. If a CBSDs has no grants then the CBSD is removed from the list.
+
+  Args:
+    cbsds: List of CbsdData objects that are in the neighbourhood to an FSS.
```
You   are correct. CbsdData object is of the format defined in section8.3 of   WINNF-TS-0096.           The json schema for this data type is actually defined under the PR#296   (FAD.1). It has been approved and should get merged soon. I think the schema   definition and example given under this PR should be sufficient. The schema   is defined in   "Spectrum-Access-System/schema/CbsdData.schema.json".          Hope this is ok.

sbdt-google replied:
Ah ok, if it become part of the schema with another PR soon, it should be fine.

jni2000 commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,357 @@
+
+  # Update sfss and pfss list. pfss contains the list of CBSDs for which grants needs to be purged.
+  if len(sfss) != 0:
+    pfss = sfss[nc:]
```
Updated pfss to   cbsds_to_purge

jni2000 commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,357 @@
+
+  # Find the largest index (nc) such that aggregation of the interferences do not cross 
+  # the OOBE threshold
+  for index, cbsd in enumerate(sfss):
```
Thanks   for simplified code :). It has been updated accordingly.

jni2000 commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,357 @@
+        neighbouring_cbsds.append(cbsd)
+  return neighbouring_cbsds
+
+def dbmAddition(value1, value2):
```
Yes. It   has been removed.

jni2000 commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,357 @@
+      for index, cbsd in enumerate(pfss):
+        # Get the grant that was used for the OOBE interference calculation, that
+        # is the grant with highest highFrequency.
+        grant_with_max_high_freq = max(cbsd['grantRequests'],\
```
Modified as   suggested.

jni2000 commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,357 @@
+
+          # Adding the -13 emission mask grants to the purge list 
+          
+          grants_to_purge = [grant_request for grant_request in cbsd['grantRequests'] \
```
Modified as   suggested.

jni2000 commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,357 @@
+       return
+    else:
+      for index, cbsd in enumerate(pfss):
+        # Get the grant that was used for the OOBE interference calculation, that
```
Sorry.   Didn't quite catch this comment.

sbdt-google replied:
Sorry, I was saying that I reviewed only partially the business logic of the code below (i did not checked carefully the reference documents)

jni2000 commented:
Serge, thanks for reviewing and giving comments. We have addressed all but two of your comments. The two remaining ones are
1. AMSL to AGL conversion: we will discuss this in the weekly working group meeting and change accordingly when answer is available.
2. Can you be more specific on "Note: only partial comments below on the business logic" comment to line 169 of src/harness/reference_models/fss_purge/fss_purge.py ?

Thanks and please continue review.

sbdt-google commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,333 @@
+def convertAmslToAgl(latitude, longitude, height_amsl):
+  """ Convert AMSL height to AGL height."""
+
+  altitude_cbsd = terrainDriver.GetTerrainElevation(latitude, longitude)
```
This can still not work, terrainDriver is not define in this workspace.
Use instead: ```wf_itm.terrainDriver.Get...```

Note that this should have been captured if you add some basic unit test. I would recommend that you add some unit test, in the same fashion that we find in other parts of the reference_models. 
Basically a file named fss_purge_test.py, which try to unit test your interface on some simple scenarios.
To be independent of the propag results, we need to mock the propagation routines. 
One simple method is to introduce a 'fake' ITM propagation model that returns a pathloss equal to the distance + an offset.  Hence we can build scenarios, and test basic functioning.
At the very least we should have an 'example' like was done for the DPA and PPA reference models, which exercise the code, and we rely on internal testing performed by the various companies.

jni2000 replied:
Thanks. The terrain driver is working after making the change you suggested. We had some testing done on this model but all the height values were in AGL format hence this piece of code got missed. 

We have added new ‘testdata’ directory containing CbsdData objects used for unit testing of FSS Purge Model. A new file containing the unittest code has also been added. 

Now we have made the changes and tested it. We have tested the model by exercising the code similar to DPA and PPA reference models. Mocking was required at many places for db_loss, incidence_angles, antenna_gain etc.  As the main function of this model is to ensure that the sorted list (sfss) and remove list (pfss) are handled correctly and the grants are removed properly. We didn’t go for ‘mock’ approach. We fine-tuned the test data in order to hit the below cases
1.	More than one iteration is done
2.	A CBSD has some grant removed and some grants remaining
3.	A CBSD has all grants removed and hence is removed from the input FAD objects list
4.	A CBSD in pfss gets added to sfss list and causes another CBSD to move into pfss list.

Test result
Input list:  cbsd_1_grant_1,cbsd_1_grant_2,cbsd_1_grant_3,cbsd_2_grant_1,cbsd_2_grant_2,cbsd_2_grant_3,cbsd_3_grant_1,cbsd_3_grant_2,cbsd_3_grant_3,cbsd_4_grant_1,cbsd_5_grant_1,cbsd_6_grant_1
Output list: cbsd_1_grant_1,cbsd_1_grant_2,cbsd_3_grant_1,cbsd_3_grant_2,cbsd_3_grant_3,cbsd_4_grant_1,cbsd_5_grant_1,cbsd_6_grant_1

sbdt-google replied:
Thanks for adding some unit test. This is a good direction to take as it will give us confidence we do not introduce regression.
Regarding the approach for unit testing and mocking, I have just proposed a FakePropagation class in another PR #476, which allow for having a controlled propagation, so that unit test  are more easy to create (for example please check the ppa_test.py file I created on that PR). 
I will continue to add utility libraries as I see fit to help doing such unit testing.

jni2000 commented:
Serge, we have addressed you latest review comments and added unit test. Please continue review. Thanks.

sbdt-google commented:
Thanks, I will try having someone else review the pure business logic soon, and kate for the interfacing to the test harness.

kate-harrison commented on src/harness/reference_models/fss_purge/testdata/fss_record_0.json:
```diff
@@ -0,0 +1,31 @@
+          "highFrequency": 4200000000
+        }
+      },
+      "ttc": true,
```
This (and the two parameters below) is not part of the SAS Testing API. Do we need to add it?

jni2000 replied:
These   two parameters have been added as part of this submission into the fss   records.

kate-harrison replied:
Please propose this change to the testing API via the test dev mailing list so that everyone is aware and can comment.

kate-harrison replied:
You will also need to update the existing FSS test data. It's probably best to do that in a separate PR so that it can get submitted quickly.

jni2000 replied:
Agreed. We have submitted a new PR #486 for this. Please review.

kate-harrison commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,332 @@
+  # Convert the height to AGL if it was AMSL type.
+  if fss_height_type == 'AMSL':
+     fss_height_agl = convertAmslToAgl(fss_latitude, fss_longitude, fss_height)
+  else:
```
You might want to throw an error if the value isn't 'AGL'.

jni2000 replied:
Updated   the code to throw an exception if its unsupported height type

kate-harrison commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,332 @@
+                                  weight_2=fss_weight_2)
+  return fss_entity
+
+def getNeighbouringCbsdsToFss(cbsd_list, fss_record):
```
Please use American spellings.

jni2000 replied:
Updated   neighbouring to neighboring

kate-harrison commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,332 @@
+  """ Get the list of all CBSDs that lie within 40 KMs of FSS.
+
+  Args:
+    cbsd_list: List of CBSDs.
```
Registration requests? CbsdData dictionaries? CBSD objects? Something else?

jni2000 replied:
It is   CbsdData dictionary. The text has been updated.

kate-harrison commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,332 @@
+  neighbouring_cbsds = []
+  for cbsd in cbsd_list:
+    distance_km, _, _ = vincenty.GeodesicDistanceBearing(
+        fss_record['deploymentParam'][0]['installationParam']['latitude'],
```
Why only the first?

jni2000 replied:
The   installation parameter is defined as a list and I think that even if multiple   items present they should share the same latitude and longitude.

kate-harrison replied:
I'm not sure it's safe to assume this. The test could be misconfigured. It probably makes sense to double-check this assumption somewhere (maybe with a helper function which is called in both MCP.1 and at the beginning of the purge list?). At least document the assumption.

kate-harrison commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,332 @@
+  """
+
+  # Calculate the OOBE interference value for each CBSD based on the grant with the highest frequency.
+  calculateOobeInterference(cbsds, fss_entity)
```
Are you sure it's safe to modify the input?

jni2000 replied:
The   'oobe_interference' parameter that is added to the input is removed at the   end of this function. The addition and removal of this parameter is done   within the performPurge method. This parameter makes the algorithm easier to   implement without the need for an additional list and mapping to the input.   We think it should be ok because the parameter is finally removed and the   input values remain the same before the fss purge model is invoked. Please   let us know if you would like logic to be changed.

kate-harrison commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,332 @@
+  Args:
+    cbsds: List of CbsdData objects that are in the neighbourhood to an FSS.
+    fss_entity: The FssProtectionPoint tuple of the FSS for which the purge is to be performed.
+    fss_oobe_margin: OOBE Margin value to be considered for FSS purge list calculation.
```
This should be defined at the top of the file, not configurable.

jni2000 replied:
Created  a    variable named FSS_OOBE_MARGIN at the top of the file.

kate-harrison commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,332 @@
+    sas_test_harness_fads: A list of FullActivityDump objects containing the FAD records 
+      from SAS test harnesses.
+    fss_record: A single FSS record dictionary.
+    fss_oobe_margin: OOBE Margin value to be considered for FSS purge list calculation.
```
Not a parameter.

jni2000 replied:
Updated   accordingly

kate-harrison commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,332 @@
+  """
+
+  # If the FSS is of TT&C type then perform the FSS purge model for the FSS.
+  if fss_record['deploymentParam'][0]['ttc']:
```
This parameter is not currently part of an FSS record.

jni2000 replied:
FSS record has been updated with this field   as part of this PR.

kate-harrison commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,332 @@
+      fss_entity = getFssInfo(fss_record)
+      performPurge(neighbouring_cbsds, fss_entity, fss_oobe_margin)
+
+  return 
```
What are you doing with the output??

jni2000 replied:
Purging   the grants is directly done to the input parameters. Hence no output is   returned.

kate-harrison commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,332 @@
+    for fad in sas_test_harness_fads:
+      cbsds.extend(fad.getCbsdRecords())
+
+    neighbouring_cbsds = getNeighbouringCbsdsToFss(cbsds, fss_record)
```
American spelling, please.

jni2000 replied:
Updated   neighbouring to neighboring

jni2000 replied:
Updated   neighbouring to neighboring

kate-harrison commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,332 @@
+
+# Define FSS information tuple containing the parameters necessary for antenna
+# and propagation calculation as listed in the test spec
+FssProtectionPoint = namedtuple('FssProtectionPoint',
```
This looks like it was copied from another location. Should we extract that into a common library instead of copying?

jni2000 replied:
This   tuple is also used in the IAP reference model and is define din the   'interference' model. Interference model is already made a common library for   the 'aggregate_interfernce' and 'IAP' models. Hence we also think it makes   sense to use this tuple definition from the interference model rather than   defining a local copy here.

kate-harrison commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,332 @@
+      for grant_items in grants_to_purge:
+         cbsd['grantRequests'].remove(grant_items)
+
+      # If no grants are remaining then remove the CBSD.
```
Why does the CBSD need to deregister after losing its Grants?

jni2000 replied:
Code has   been updated to retain CBSDs even if all its grants are removed.

kate-harrison commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,332 @@
+
+        # Adding the -13 emission mask grants to the purge list 
+        
+        grants_to_purge = [grant_request for grant_request in cbsd['grantRequests'] 
```
If `cbsd` is a `CbsdData` object from SAS-SAS, then you need to use `grants` not `grantRequests` here.

jni2000 replied:
Updated

kate-harrison commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,332 @@
+  """
+
+  # Calculate the OOBE interference value for each CBSD based on the grant with the highest frequency.
+  calculateOobeInterference(cbsds, fss_entity)
```
Does this need to be done every time you enter this function? Since this implementation is recursive, that seems wasteful unless you expect the value to change over time.

jni2000 replied:
Actually   the value might change for each iteration. This method   'calculateOobeInterference' uses only the grant with the highest value of   'highfrequency'. If this grant that was used for oobe_interference   calcualation is removed by the algorithm then in the next iteration a   different grant will be used for the calculation.

kate-harrison commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,332 @@
+
+      # If no grants are remaining then remove the CBSD.
+      if not cbsd['grantRequests']:
+         cbsds_to_purge.remove(cbsd)
```
What do you ultimately do with `cbsds_to_purge`? It seems to be unused.

jni2000 replied:
Ultimately   the cbsds_to_purge should become empty. The algorithm runs in recursively   until there are no elements in the cbsds_to_purge list.

kate-harrison commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,332 @@
+    
+  # Removing the oobe_interference parameter added to cbsd objects
+  for cbsd in cbsds:
+    cbsd.pop('oobe_interference', None)
```
You can use `del` instead since you don't need to retrieve the value.

jni2000 replied:
Updated   with del function

jni2000 commented:
Kate, we have addressed your recent review comments. Please continue. Also please note that  in our current implementation we modified the input FADs directly. We will change to explicit dedicated output FADs all together for all the purging reference modules in next drop.

nasrinivasa commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,310 @@
+    neighboring_cbsds = getNeighboringCbsdsToFss(cbsds, fss_record)
+    if neighboring_cbsds:
+      fss_entity = getFssInfo(fss_record)
+      performPurge(neighboring_cbsds, fss_entity)
```
Shouldn't the results of the Purge List module (list of CBSDs that are purged) be available for use in the pre-IAP filtering of CBSDs? The fssPurgemodel and the performPurge does not seem to return the values.

jni2000 replied:
The original approach was to make change to the input fad objects. The approach has been changed now. Each reference model will not modify the input FAD objects and they will return FAD objects with the modified/purged content.

Now fssPurgeModel will return FAD objects. The performPurge function does not provide any return values as it is an internal function to the FSS purge model and it modifies the input parameters.

nasrinivasa commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,310 @@
+REF_BW = 5.e6
+
+# OOBE Margin
+FSS_OOBE_MARGIN = 1
```
Shouldn't the Margin value be moved out into a configuration file rather than setting it here

jni2000 replied:
In an earlier comment Kate   mentioned these margin values will not be configurable. Hence they are   hardcoded in the reference model.

nasrinivasa replied:
Will it be part of the input configuration for the test cases, meaning will it be possible for the margin value to be set for a given test case and changed for another input configuration? If that is true then it will be helpful if they are configured from outside

jni2000 replied:
Agreed with your suggestion. But Kate has a different opinion on this aiming to simplify the test cases avoiding operation time mis-configuration caused errors.

kate-harrison replied:
It seems like now is the right time for me to chime in on this :) WinnForum's position is that the margin values are not configurable within the test infrastructure. The margin values are specified via a steering committee policy document [1] and the test code should use exactly those values. I strongly support the idea of having a figurative #define at the top of this file, but I don't think we should go any further than that. (And to be clear, my concern is less about mis-configuration potential and more a matter of matching the WF specs and policies.)

[1] https://workspace.winnforum.org/higherlogic/ws/public/download/5676/WINNF-SSC-0007-V1.0.1%20-%20WinnForum%20CBRS%20Protection%20Margin%20Policy.pdf

jni2000 replied:
Thanks for clarifying. I only remembered a portion and the least important portion of the reasons :)

nasrinivasa replied:
Thanks for clarifying this Kate

nasrinivasa commented on src/harness/reference_models/fss_purge/testdata/fss_record_0.json:
```diff
@@ -0,0 +1,29 @@
+        "antennaBeamwidth": 0,
+        "antennaAzimuth": 270,
+        "antennaDowntilt": 0,
+        "indoorDeployment": true,
```
The indoorDeployment flag should not be part of the fss_record

jni2000 replied:
Thanks for catching this. It has   been removed.

nasrinivasa commented on src/harness/reference_models/fss_purge/testdata/cbsd_1.json:
```diff
@@ -0,0 +1,71 @@
+                                                "operationFrequencyRange": {
+                                                        "highFrequency": 3699000000.0,
+                                                        "lowFrequency": 3696000000.0
+                                                }
```
Multiple grants cannot have overlapping frequency bands

jni2000 replied:
changed.

jni2000 commented:
Kate and Navin

We have addressed your review comments, please continue review. Thanks.

nasrinivasa commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,318 @@
+        # Get all the grants that share the same emission mask as the one that 
+        # was used for OOBE calculation
+        if (grant_with_max_high_freq['operationParam']['operationFrequencyRange']
+            ['highFrequency'] > 3690000000):
```
The logic here assumes that the FSS band type for the TT&C type starts at 3700. This will be controlled by the input configuration.Do we need to have some code that would validate the input and ensure that the inputs conform to this assumption? Otherwise, the logic here would not work.

jni2000 replied:
As per update from WinnF WG.4,   purge list will be only calculated for 3700MHz -3705 MHz (for the FSS earth   station with TT&C designation).

TaranFurnell commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,300 @@
+
+        # Removing grant requests that are in the purge list from the CBSDs.
+        for grant_items in grants_to_purge:
+           cbsd['grants'].remove(grant_items)
```
Each FSS must be considered independently for all grants. Here the grants are being removed from the input which can result in FSS purge list results being dependent on the order the FSSs are provided in. Please refactor this to prevent the result being FSS order dependent.

jni2000 replied:
Good point about grants interfering with more than one FSS. However, I think that the dependency has to be taken into account and the grants removed for one FSS should not be considered for another FSS. A grant removed for one FSS should not be present in the final list and I think it should also not be considered for another FSS because considering that grant might cause some other valid grant to get purged.

For example consider two FSSs F1, F2 and six Grants G1, G2, G3, G4, G5 and G6. If the grants_to_purge for ‘F1’ is [G4, G5] and if these two grants are considered for F2 and the sorted list of grants_to_keep becomes [G4,G5, G3, G1] and the grants_to_purge for F2 becomes [G2, G6] then grant G2 and G6 would get purged. If we had not considered the grants G4 and G5 then perhaps grant G2 or both grants G2 and G6 might have ended in grants_to_keep list and would not get purged.

I completely agree on the point that the order of FSSs should not be a dependency. I propose an approach wherein the FSS purge model for all FSS records is executed recursively if the previous run resulted in any grants getting purged for any of the FSS records. A high level psedo-code for the approach is below.

fssPurgeReferenceModel (uut_fad, sas_th_fads, fss_records)
  set cbsd_list to the list of all cbsds from the  fad objects of UUT and THs
  re_run_flag = False
  for each fss_record in fss_records
    no_of_grants_purged = performPurge(cbsd_list, fss_record)
    if no_of_grants_purged > 0
      re_run_flag = True

  if re_run_flag:
    fssPurgeReferenceModel(uut_fad, sas_th_fads, fss_records)

TaranFurnell replied:
This doesn't match the spec. I suggest you talk to Masoud and Sam. After that if you still think it is appropriate please bring it up in the Tuesday call.

jni2000 replied:
Taran

After discussing with Sam, we believe what you suggested is correct and thus we make the changes to reflect this.

jni2000 commented:
Taran, we reviewed your last comment and presented out thoughts. We came out a solution to address your concern, please take a look.

nasrinivasa commented on src/harness/reference_models/fss_purge/testdata/cbsd_2.json:
```diff
@@ -0,0 +1,71 @@
+					"operationParam": {
+						"maxEirp": 37,
+						"operationFrequencyRange": {
+							"highFrequency": 3670000000.0,
```
Frequency needs to be an integer

jni2000 replied:
fixed

nasrinivasa commented on src/harness/reference_models/fss_purge/testdata/cbsd_2.json:
```diff
@@ -0,0 +1,71 @@
+                                                "maxEirp": 37,
+                                                "operationFrequencyRange": {
+                                                        "highFrequency": 3680000000.0,
+                                                        "lowFrequency": 3670000000.0
```
Frequency needs to be an integer

jni2000 replied:
fixed

nasrinivasa commented on src/harness/reference_models/fss_purge/testdata/cbsd_2.json:
```diff
@@ -0,0 +1,71 @@
+                                                "maxEirp": 37,
+                                                "operationFrequencyRange": {
+                                                        "highFrequency": 3699000000.0,
+                                                        "lowFrequency": 3696000000.0
```
Frequency needs to be an integer

jni2000 replied:
fixed

nasrinivasa commented on src/harness/reference_models/fss_purge/testdata/cbsd_2.json:
```diff
@@ -0,0 +1,71 @@
+					"grantExpireTime": "calculated at runtime",
+					"id": "cbsd_2_grant_0",
+					"operationParam": {
+						"maxEirp": 37,
```
maxEirp should be with decimal points

jni2000 replied:
fixed

nasrinivasa commented on src/harness/reference_models/fss_purge/testdata/cbsd_2.json:
```diff
@@ -0,0 +1,71 @@
+                                        "grantExpireTime": "calculated at runtime",
+                                        "id": "cbsd_2_grant_1",
+                                        "operationParam": {
+                                                "maxEirp": 37,
```
maxEirp should be with decimal points

jni2000 replied:
fixed

nasrinivasa commented on src/harness/reference_models/fss_purge/testdata/cbsd_2.json:
```diff
@@ -0,0 +1,71 @@
+                                        "grantExpireTime": "calculated at runtime",
+                                        "id": "cbsd_2_grant_2",
+                                        "operationParam": {
+                                                "maxEirp": 37,
```
maxEirp should be with decimal points

jni2000 replied:
fixed

nasrinivasa commented on src/harness/reference_models/fss_purge/testdata/cbsd_3.json:
```diff
@@ -0,0 +1,45 @@
+					"grantExpireTime": "calculated at runtime",
+					"id": "cbsd_3_grant_0",
+					"operationParam": {
+						"maxEirp": 37,
```
maxEirp should be with decimal points

jni2000 replied:
fixed

nasrinivasa commented on src/harness/reference_models/fss_purge/testdata/cbsd_3.json:
```diff
@@ -0,0 +1,45 @@
+						"maxEirp": 37,
+						"operationFrequencyRange": {
+							"highFrequency": 3640000000.0,
+							"lowFrequency": 3630000000.0
```
Frequency should be integer

jni2000 replied:
fixed.

nasrinivasa commented on src/harness/reference_models/fss_purge/testdata/cbsd_4.json:
```diff
@@ -0,0 +1,45 @@
+					"grantExpireTime": "calculated at runtime",
+					"id": "cbsd_4_grant_0",
+					"operationParam": {
+						"maxEirp": 37,
```
maxEirp should be with decimal points

jni2000 replied:
fixed

nasrinivasa commented on src/harness/reference_models/fss_purge/testdata/cbsd_4.json:
```diff
@@ -0,0 +1,45 @@
+						"maxEirp": 37,
+						"operationFrequencyRange": {
+							"highFrequency": 3690000000.0,
+							"lowFrequency": 3680000000.0
```
Frequency should be integer

nasrinivasa commented on src/harness/reference_models/fss_purge/testdata/cbsd_5.json:
```diff
@@ -0,0 +1,45 @@
+						"operationFrequencyRange": {
+							"highFrequency": 3700000000.0,
+							"lowFrequency": 3690000000.0
+						}
```
maxEirp should be with decimal points
Frequency should be integer

jni2000 replied:
fixed

nasrinivasa commented on src/harness/reference_models/fss_purge/testdata/cbsd_1.json:
```diff
@@ -0,0 +1,71 @@
+                                                "operationFrequencyRange": {
+                                                        "highFrequency": 3698000000.0,
+                                                        "lowFrequency": 3696000000.0
+                                                }
```
maxEirp should be with decimal points
Frequency should be integer

jni2000 replied:
fixed

nasrinivasa commented on src/harness/reference_models/fss_purge/testdata/cbsd_0.json:
```diff
@@ -0,0 +1,71 @@
+                                                "operationFrequencyRange": {
+                                                        "highFrequency": 3695000000.0,
+                                                        "lowFrequency": 3690000000.0
+                                                }
```
maxEirp should be with decimal points
Frequency should be integer

jni2000 replied:
fixed

nasrinivasa commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,300 @@
+  fss_pointing_elevation = fss_record['deploymentParam'][0]['installationParam']['elevation']
+  fss_max_gain_dbi = fss_record['deploymentParam'][0]['installationParam']['antennaGain']
+  fss_weight_1 = fss_record['deploymentParam'][0]['weight1']
+  fss_weight_2 = fss_record['deploymentParam'][0]['weight2']
```
fss_weight_1 and fss_weight_2 are not part of the FSS record

jni2000 replied:
removed

nasrinivasa commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,300 @@
+"""
+import numpy as np
+from reference_models.antenna import antenna
+from reference_models.interference import interference
```
Is interference.py part of this pull request?

jni2000 replied:
This used to be included in this PR, but move out as separate one per review comments. I'll check this out where it goes to and get back to you on this. Thanks.

jni2000 commented:
Navin

We have address all your latest review comments and made necessary corrections as you suggested. Thanks. Please take a look.

Taran

For your last comment, I will discuss thoroughly with Sam and get back to you tomorrow. Thanks.

TaranFurnell commented:
Hi @jni2000, have you had a chance to talk to Sam yet?

jni2000 commented:
Taran, sorry for late response. We have discussed your below comment and agreed with you that it is safer to do as you said. We are working on this fix now and will update asap.

jni2000 commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,271 @@
+#    Copyright 2018 SAS Project Authors. All Rights Reserved.
```
done as suggested.

jni2000 commented:
Serge, we have moved the codes to pre_iap_filtering directory as you suggested.

Taran, we also made changes to correct the handling logic as you commented below:

> Each FSS must be considered independently for all grants. Here the grants are being removed from the input which can result in FSS purge list results being dependent on the order the FSSs are provided in. Please refactor this to prevent the result being FSS order dependent.

Please review.

TaranFurnell commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,296 @@
+                                               pointing_elevation=fss_pointing_elevation)
+  return fss_entity
+
+def performPurge(cbsds, fss_entity):
```
See comment on line 280.

jni2000 replied:
Re-modified   the code logic

TaranFurnell commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,296 @@
+
+          # Setting the purge_flag for the grants that need to be purged
+          for grant_items in grants_to_purge_for_cbsd:
+             for grant in cbsd['grants']:
```
This line has inconsistent spacing, please use 2 spaces.

jni2000 replied:
Modified   the spaces to be consistent across the file

sbdt-google commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,296 @@
+from reference_models.interference import interference
+from reference_models.geo import vincenty
+from reference_models.propagation import wf_itm
+import pre_iap_util
```
Please use the full path:
```
from reference_models.pre_iap_filtering import pre_iap_util
```
Relative path are not robust on the python configuration setup.

jni2000 replied:
Modified   to absolute path

sbdt-google commented on src/harness/reference_models/pre_iap_filtering/fss_purge_test.py:
```diff
@@ -0,0 +1,73 @@
+import os
+import unittest
+import full_activity_dump
+import fss_purge
```
Also here, please put absolute import from the harness/ directory (which is the root python path):
```
from reference_models.pre_iap_filtering import fss_purge
```

jni2000 replied:
Modified   to absolute path

sbdt-google commented on src/harness/reference_models/pre_iap_filtering/fss_purge_test.py:
```diff
@@ -0,0 +1,73 @@
+   sas_uut_fad_object = fad_object_1
+   sas_test_harness_fad_object = [fad_object_2, fad_object_3]
+ 
+   print "================CBSD Grants passed as input======================"
```
This looks like a unit test but is not really one, until you assert expected output with the self.assertXXX functions.
So that we can detect regressions.
Can you add such tests ? Or do you want to do as part of another PR ?

nasrinivasa commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,296 @@
+      if neighboring_cbsds_with_grants:
+        fss_entity = getFssInfo(fss_record)
+        grants_to_purge_for_fss = performPurge(neighboring_cbsds_with_grants, fss_entity)
+        grants_to_purged_for_all_fss.extend(grants_to_purge_for_fss)
```
Duplicate grant entries (due to a grant being common to multiple FSS) are not removed from the grants_to_purged_for_fss list. This could cause the list to become huge. Shouldn't the duplicates be pruned before using the list for futher processing?

jni2000 replied:
As   suggested, modified the code to not create duplicate entries

nasrinivasa commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,296 @@
+    # Sort the CBSDs based on the calculated interference value.
+    sorted_cbsds = sorted(cbsds, key=lambda cbsd: cbsd['oobe_interference'])
+    # Calculate the threshold value in dBm for the OOBE interference
+    oobe_threshold_value = -129 + 10 * (np.log10(REF_BW/interference.ONE_MHZ) - FSS_OOBE_MARGIN)
```
oobe_threshold_value is a constant, so can be moved out of the while loop such that we don't calculate it always. It could even be set once and be used for all FSS protection points of interest

jni2000 replied:
As   suggested, moved "oobe_threshold_value" calculation outside while   loop

nasrinivasa commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,296 @@
+          for grant_items in grants_to_purge_for_cbsd:
+             for grant in cbsd['grants']:
+               if grant == grant_items:
+                 grant['purge_flag'] = True
```
Does this line update the "cbsds" list with the purge_flag? I don't see any place where the cbsds list is updated in the while loop. All the operations are done based on the sorted_cbsds list which is a sorted copy of the cbsds list.

jni2000 replied:
Re-modified   the code logic

nasrinivasa commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,296 @@
+  fss_height = fss_record['record']['deploymentParam'][0]['installationParam']['height']
+  fss_height_type = fss_record['record']['deploymentParam'][0]['installationParam']['heightType']
+  fss_azimuth = fss_record['record']['deploymentParam'][0]['installationParam']['antennaAzimuth']
+  fss_pointing_elevation = fss_record['record']['deploymentParam'][0]['installationParam']['elevation']
```
Should be extracted using the antennaDowntilt key and not the elevation

jni2000 replied:
Modified   key as "antennaDowntilt" for fss_pointing_elevation calculation

jni2000 commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,296 @@
+      gcbsd = getAntennaGainTowardsFss(incidence_angle, cbsd)
+      gfss = getAntennaGainFssReceiver(fss_entity, incidence_angle)
+      oobe_interference = mcbsd + gcbsd - lcbsd + gfss
+      cbsd['oobe_interference'] = oobe_interference
```
Re-modified   the code logic

jni2000 commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,296 @@
+  # Removing the oobe_interference and purge_flag keys added to CBSD objects
+  for cbsd in cbsds:
+    if 'oobe_interference' in cbsd:
+      del cbsd['oobe_interference']
```
Re-modified   the code logic

jni2000 commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,296 @@
+  for cbsd in cbsds:
+      for grant in grants_to_purged_for_all_fss:
+        if grant in cbsd['grants']:
+          cbsd['grants'].remove(grant)
```
Modified   the spaces to be consistent across the file

jni2000 commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,296 @@
+  if cbsd_height_type == 'AMSL':
+      cbsd_height_agl = convertAmslToAgl(cbsd_latitude, cbsd_longitude, cbsd_height)
+  else:
+      cbsd_height_agl = cbsd_height
```
Modified   the spaces to be consistent across the file

jni2000 commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,296 @@
+  elif fss_height_type == 'AGL':
+     fss_height_agl = fss_height
+  else:
+     raise Exception('Invalid FSS height type.')
```
Modified   the spaces to be consistent across the file

jni2000 commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,296 @@
+          for grant_items in grants_to_purge_for_cbsd:
+             for grant in cbsd['grants']:
+               if grant == grant_items:
+                 grant['purge_flag'] = True
```
Re-modified   the code logic

jni2000 commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,296 @@
+
+  # Removing grant requests that are in the grants purge list from the CBSDs.
+  for cbsd in cbsds:
+      for grant in grants_to_purged_for_all_fss:
```
As   suggested, modified the code to have key, values pair and purging grant with   cbsd key

jni2000 commented on src/harness/reference_models/fss_purge/fss_purge.py:
```diff
@@ -0,0 +1,271 @@
+  fss_height = fss_record['record']['deploymentParam'][0]['installationParam']['height']
+  fss_height_type = fss_record['record']['deploymentParam'][0]['installationParam']['heightType']
+  fss_azimuth = fss_record['record']['deploymentParam'][0]['installationParam']['antennaAzimuth']
+  fss_pointing_elevation = fss_record['record']['deploymentParam'][0]['installationParam']['elevation']
```
Modified   as suggested

jni2000 commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,296 @@
+                                               pointing_elevation=fss_pointing_elevation)
+  return fss_entity
+
+def performPurge(cbsds, fss_entity):
```
Updated   the function document

jni2000 commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,296 @@
+    # Sort the CBSDs based on the calculated interference value.
+    sorted_cbsds = sorted(cbsds, key=lambda cbsd: cbsd['oobe_interference'])
+    # Calculate the threshold value in dBm for the OOBE interference
+    oobe_threshold_value = -129 + 10 * (np.log10(REF_BW/interference.ONE_MHZ) - FSS_OOBE_MARGIN)
```
we are studying this now, will response soon.

TaranFurnell commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,301 @@
+  return fss_ant_gain
+
+def getHighestGrantOfCbsds(cbsds):
+  """Creates a new list of grants for oobe_calculation
```
This doesn't really provide a list of grants.

jni2000 replied:
Modified   as per new functionality

TaranFurnell commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,301 @@
+        # lower emission mask then add it to the remaining_grants list.
+        for grant in purged_grant['cbsd']['grants']:
+          # Check that we are not comapring the same grant as the pruged grant
+          if grant != purged_grant['grantData']:
```
This is creating a special case without a good reason.

jni2000 replied:
In this   function we wanted to identify other grants of cbsd with the same emission   mask, extra check is introduced to avoid grant to compare with same(self)   grant

TaranFurnell replied:
My point is there is no need to do this check. Removing the bit at the start (now lines 110-111) that adds all the directly referenced grants to ```final_purge_list``` and then removing this check will make the code easier to read while doing exactly the same thing.

TaranFurnell commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,301 @@
+  else:
+      mcbsd = -25
+
+  return mcbsd
```
Can be one-lined:

```python
return -13 if grant['operationParam']['operationFrequencyRange']['highFrequency'] > 3690000000 else -25
```

jni2000 replied:
Modified   to one-line

jni2000 commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,301 @@
+                                               pointing_elevation=fss_pointing_elevation)
+  return fss_entity
+
+def performPurge(cbsds, fss_entity):
```
Modified   the name to "generatePurgeListForFssPoint"

jni2000 commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,301 @@
+        fss_entity = getFssInfo(fss_record)
+        grants_to_purge_for_fss = performPurge(neighboring_cbsds_with_grants, fss_entity)
+        grants_to_purged_for_all_fss.extend(purge_data for purge_data in grants_to_purge_for_fss
+                                                     if purge_data not in grants_to_purged_for_all_fss)
```
We   have modified the "perforumPurge" to return named tuples now, since   it returns dictionaries we are not able to do set operations on them, so kept   the code as it is, we are working on it to avoid duplicate entries with   efficient way.

TaranFurnell replied:
I don't understand what you mean by this comment. Named tuples can be added to a set, they are not dictionaries.

jni2000 commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,301 @@
+  final_purge_list = []
+
+  # List of dictionaries containing the highest grant of the CBSD and a reference to the CBSD.
+  grants_for_oobe_calculation = getHighestGrantOfCbsds(cbsds)
```
Agreed.Modified   to "grants_cbsds_info_for_oobe_calculation"

jni2000 commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,301 @@
+  The interference values are calculated based on the grant with the highest highFrequency value.
+  The calculated value is added into the grant object.
+  Args:
+    grants: List of dictionaries contianing the highest grant of the CBSD and a reference to the CBSD.
```
Modified   to "grants_cbsds_oobe_info" at appropriate places

TaranFurnell replied:
Please update the comment.

jni2000 commented:
Taran, thanks for the review comments. We have addressed them, please take a look.

TaranFurnell commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -115,25 +120,25 @@ def performPurge(cbsds, fss_entity):
           if grant != purged_grant['grantData']:
-            if (getMcbsdValue (purged_grant['grantData']) == getMcbsdValue (grant)):
-              final_purge_list.append({'grantData': grant, 'cbsd': purged_grant['cbsd']})
+            if (getMcbsdValue(purged_grant['grantData']) == getMcbsdValue(grant)):
```
If a cbsd has grants in both emission masks this will cause an infinite loop.

TaranFurnell commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,301 @@
+    max_grant = max(cbsd['grants'],\
+                    key=lambda x: x['operationParam']['operationFrequencyRange']['highFrequency'])
+    grant_info = {
+        'grantData': max_grant,
```
Per my comment above. The part that is using the ```grantData``` directly isn't really necessary.

nasrinivasa commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,301 @@
+        if len(remaining_grants) > 0:
+          max_grant = max(remaining_grants,\
+                           key=lambda x: x['operationParam']['operationFrequencyRange']['highFrequency'])
+          grants_cbsds_info_for_oobe_calculation.append({'grantData': max_grant, 'cbsd': purged_grant['cbsd']})
```
Shouldn't we remove the purged grants before appending the new grants into grants_cbsds_info_for_oobe_calculation

TaranFurnell replied:
I think the current implementation is correct (barring previously mentioned bugs), grants_cbsds_info_for_oobe_calculation is all the CBSD-grant tuples that are not in the purgelist for this iteration. The next iteration must consider both those grants/CBSDs and any grants from CBSDs that also have a grant at a lower emission mask, which this is supposed to be adding.

nasrinivasa replied:
I see that now. Missed the line where the assignment is being done

TaranFurnell commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,301 @@
+
+        # Identify the grant with the highest highFrequncy value among the remaining grants for the
+        # CBSD and add it to the grants_cbsds_info_for_oobe_calculation list for next iteration.
+        if len(remaining_grants) > 0:
```
[] is falsy.

```python
if remaining_grants:
```

jni2000 replied:
Modified   the code, removed the len check

TaranFurnell commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,299 @@
+        fss_entity = getFssInfo(fss_record)
+        grants_to_purge_for_fss = generatePurgeListForFssPoint(neighboring_cbsds_with_grants, fss_entity)
+        grants_to_purged_for_all_fss.extend(purge_data for purge_data in grants_to_purge_for_fss
+                                                     if purge_data not in grants_to_purged_for_all_fss)
```
Since the ```generatePurgeListForFssPoint``` method now returns a list of tuples it's easy to use a set to contain the grant/cbsd tuples (Handles duplicates efficiently and cleanly).

```python
grants_to_purged_for_all_fss = set()
  ...
    ...
      grants_to_purged_for_all_fss.update(grants_to_purge_for_fss)
```

jni2000 replied:
We   tried the changes suggested in comment, but facing the error "TypeError:   unhashable type: 'dict'"     The output of "generatePurgeListForFssPoint" is a namedtuple, but   the elements 'grant' and 'cbsd' point to a dictionary, hence the set   operation fails with an error stating "TypeError: unhashable type:   'dict'".

TaranFurnell replied:
minor correction: ```'cbsdId'``` and ```'grantId'``` should both be ```'id'```.

TaranFurnell commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,299 @@
+            for final_grant in final_purge_list:
+              if cbsd_grant is final_grant.grant:
+                break
+            else:
```
There isn't a need to check the ```final_purge_list```, ignore the grant unless the grants mcbsd is lower than the mcbsd in purged_grant.

```python
if purged_grant['mcbsd'] == getMcbsdValue(cbsd_grant):
...
elif purged_grant['mcbsd'] > getMcbsdValue(cbsd_grant):
  remaining_grants.append(cbsd_grant)
```

jni2000 replied:
updated   the code as comment suggested

TaranFurnell commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,299 @@
+        remaining_grants = []
+        for cbsd_grant in purged_grant['cbsd']['grants']:
+          # Find the grants matching mcbsd value and add it to the purge list
+          if (purged_grant['mcbsd']) == getMcbsdValue(cbsd_grant):
```
nit: Unnecessary brackets.

```python
if purged_grant['mcbsd'] == getMcbsdValue(cbsd_grant):
```

jni2000 replied:
Updated,   removed the brackets

jni2000 commented:
Taran, we have addressed all your comments please continue reviewing. Thanks.

jni2000 commented:
Taran, we have addressed you very last comment from yesterday, please take a look. If no more concerns, please can you allow moving forward? Thanks.

credtech commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,301 @@
+REF_BW = 5.e6
+
+# OOBE Margin
+FSS_OOBE_MARGIN = 1
```
Shouldn't the margin be set to "2" instead of "1"? Document WINNF-SSC-007-V1.0.0 states the following:

In release 1 of WinnForum specifications, the following values are agreed among and used by all SAS Administrators.
Mg1 = Mg2 = Mg3 = Mg4 = Mg5 = Mg6= 2dB.

jni2000 replied:
The newer Document WINNF-SSC-007-V1.0.1 states
Mg1 = Mg2 = Mg3 = Mg4 = Mg5 = Mg6= 1dB.

credtech commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,301 @@
+  grants_cbsds_info_for_oobe_calculation = getGrantInfoListForFssOobeCalculation(cbsds)
+
+  # Calculate the threshold value in dBm for the OOBE interference
+  oobe_threshold_value = -129 + 10 * (np.log10(REF_BW/interference.ONE_MHZ) - FSS_OOBE_MARGIN)
```
The closing bracket seems incorrect. 
Proposed update:
oobe_threshold_value = -129 + 10 * (np.log10(REF_BW/interference.ONE_MHZ)**)** - FSS_OOBE_MARGIN

credtech commented on src/harness/reference_models/pre_iap_filtering/fss_purge.py:
```diff
@@ -0,0 +1,301 @@
+
+    # Sort the grants based on the calculated interference value.
+    sorted_grants_cbsds_info = sorted(grants_cbsds_info_for_oobe_calculation, key=lambda grant: grant['oobe_interference'])
+    # Find the largest index such that aggregation of the interferences do not cross
```
According to the code below, in case aggregation is **equal** (or higher) to the treshold, the grants will be purged.
Therefore ,this line of comment is a bit misleading. 

Proposals (either is ok)

1) correct the comment in line 103 to :
# Find the largest index such that aggregation of the interferences **is not equal or higher** than the OOBE threshold.

2) Or, correct the code so that in case aggregation is **equal** to the treshold, the grants are **not** purged.

Note: in DPA reference model, option 2) is followed

jni2000 commented:
credtech

We have addressed all your three review comments. Please take a look. Thanks.

jni2000 commented:
Dear all reviewers, are we good to approve this PR now ? Thanks.

jni2000 commented:
Kate and Taran

Are you OK with this PR now ? If so, can we move on to merge ? Thanks.

credtech commented:
>We have addressed all your three review comments. Please take a look. Thanks.

Thanks for handling my comments. I'm ok with those changes.

jni2000 commented:
Kate and Taran

We got one approval of this PR, can you please take a look if we are good to move forward ? Thanks.

sbdt-google commented:
Hi James,

We are ok with the PR. If any issue discovered we can fix as part of a new
PR.
I cannot access the PR from the GitHub anymore (due to a load page error,
probably caused by the large amount of comments),
so please consider this as the approval for submit.

Thanks,
Serge

On Mon, Apr 23, 2018 at 5:49 AM, James Ni <notifications@github.com> wrote:

> Kate and Taran
>
> We got one approval of this PR, can you please take a look if we are good
> to move forward ? Thanks.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/pull/462#issuecomment-383563387>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ac5pZSYH1wWX236nxSZqreMsPuQHhZhUks5trc30gaJpZM4SjVWd>
> .
>

jni2000 commented:
Thanks, Serge. That timeout problem happened to me as well. You need to refresh multiple times (90% of time will still timeout, and I have reported the issue to github admin). Without your “approval” actually clicked, I am not allowed to merge. Can you please talk to Kate and see how we can deal with this ?

James Ni
Director of SAS System
www.federatedwireless.com<http://www.federatedwireless.com/>
https://www.linkedin.com/pub/james-ni/4/3b9/24a
978.996.7480
[Description: Macintosh HD:Users:Sam:Dropbox (Bluetext):Federated Wireless:Creative:Logo:Logo Lock Up:Final-Lockup:Federated Wireless No Tag:FedWireless-lockup-notag-web.png]

From: Serge Barbosa [mailto:notifications@github.com]
Sent: Monday, April 23, 2018 10:19 AM
To: Wireless-Innovation-Forum/Spectrum-Access-System
Cc: James Ni; Mention
Subject: Re: [Wireless-Innovation-Forum/Spectrum-Access-System] FSS purge list reference model initial delivery (#462)

Hi James,

We are ok with the PR. If any issue discovered we can fix as part of a new
PR.
I cannot access the PR from the GitHub anymore (due to a load page error,
probably caused by the large amount of comments),
so please consider this as the approval for submit.

Thanks,
Serge

On Mon, Apr 23, 2018 at 5:49 AM, James Ni <notifications@github.com<mailto:notifications@github.com>> wrote:

> Kate and Taran
>
> We got one approval of this PR, can you please take a look if we are good
> to move forward ? Thanks.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/pull/462#issuecomment-383563387>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ac5pZSYH1wWX236nxSZqreMsPuQHhZhUks5trc30gaJpZM4SjVWd>
> .
>


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/pull/462#issuecomment-383591875>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AKpZrIaIer0x49N88zq4YG6EOg0_SQY7ks5treLGgaJpZM4SjVWd>.

sbdt-google commented:
I have also reported the issue last week to GitHub, but no workaround for
now. Right now I cannot load this page anymore (nor the IAP PR), even by
trying 50 times in a row. The timeout may depends where you are located, so
you might have better luck than me.

Let me check with Kate if she can access and we will let you know soon.
Otherwise the only workaround I see, is that you you create a new PR from
your latest code revision, which is fresh of comments.

Serge

On Mon, Apr 23, 2018 at 7:31 AM, James Ni <notifications@github.com> wrote:

> Thanks, Serge. That timeout problem happened to me as well. You need to
> refresh multiple times (90% of time will still timeout, and I have reported
> the issue to github admin). Without your “approval” actually clicked, I am
> not allowed to merge. Can you please talk to Kate and see how we can deal
> with this ?
>
> James Ni
> Director of SAS System
> www.federatedwireless.com<http://www.federatedwireless.com/>
> https://www.linkedin.com/pub/james-ni/4/3b9/24a
> 978.996.7480
> [Description: Macintosh HD:Users:Sam:Dropbox (Bluetext):Federated
> Wireless:Creative:Logo:Logo Lock Up:Final-Lockup:Federated Wireless No
> Tag:FedWireless-lockup-notag-web.png]
>
> From: Serge Barbosa [mailto:notifications@github.com]
> Sent: Monday, April 23, 2018 10:19 AM
> To: Wireless-Innovation-Forum/Spectrum-Access-System
> Cc: James Ni; Mention
> Subject: Re: [Wireless-Innovation-Forum/Spectrum-Access-System] FSS purge
> list reference model initial delivery (#462)
>
> Hi James,
>
> We are ok with the PR. If any issue discovered we can fix as part of a new
> PR.
> I cannot access the PR from the GitHub anymore (due to a load page error,
> probably caused by the large amount of comments),
> so please consider this as the approval for submit.
>
> Thanks,
> Serge
>
> On Mon, Apr 23, 2018 at 5:49 AM, James Ni <notifications@github.com<
> mailto:notifications@github.com>> wrote:
>
> > Kate and Taran
> >
> > We got one approval of this PR, can you please take a look if we are
> good
> > to move forward ? Thanks.
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/Wireless-Innovation-Forum/Spectrum-
> Access-System/pull/462#issuecomment-383563387>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> Ac5pZSYH1wWX236nxSZqreMsPuQHhZhUks5trc30gaJpZM4SjVWd>
> > .
> >
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub<https://github.com/
> Wireless-Innovation-Forum/Spectrum-Access-System/pull/
> 462#issuecomment-383591875>, or mute the thread<https://github.com/
> notifications/unsubscribe-auth/AKpZrIaIer0x49N88zq4YG6EOg0_
> SQY7ks5treLGgaJpZM4SjVWd>.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/pull/462#issuecomment-383596534>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Ac5pZVO3j-jQ2BHQkB_DpRKlEvTNiIrIks5treXdgaJpZM4SjVWd>
> .
>
